rdm/base/architecture_design_chart.md

```diff
@@ -1,3 +1,6 @@
+{%- auto_section_numbering -%}
+{%- if system.auditor_notes %}{% audit_notes %}{% endif -%}
+{%- vocabulary -%}
```

Comment:
During our earlier conversation, we said:

- Including extensions in `system.yml` should only provide the extension's functionality, but not actually use it.
- Template tags will be used to provide an extension's functionality.

I think this logic makes sense most of the time, but in the case of the `VocabularyExtension`, it seems like it adds unnecessary complexity.

In particular, it feels more natural (and simpler) if adding the `VocabularyExtension` in system.yml would automatically extend the `first_pass_output` object.  This approach is nice because it won't require us to add `{%- vocabulary -%}` to each template. 

Can you think of any downsides to this approach?

Reply:
Here a couple of constraints we should stick with:
A) We use the same `system.yml` for all the documents in a set. We can use different `system.yml` files for differing outputs formats ('docx' versus 'pdf') or different audiences (auditor versus managers). But `software_plan` and `risk_analysis` use the same `system.yml`
B) When different documents need different jinja or other behavior, that must manifest as differences in the documents themselves.

With those constraints in mind we have two options:
CURRENT: `system.yml` dynamically loads the union of all the extensions needed by any of the documents. This can be different for different outputs or different audiences. If documents require an extension, they must invoke a tag defined for the extension. There can be a default behavior of the extension which is obtained by invoking the tag with no parameters. An extension can be customized for a document by adding some parameters to the invocation. Tags should be sufficiently distinct to not collide. Any context variables (such as `words` or `has`) should be sufficiently distinct to not collide. Extensions providing alternate behavior may deliberately use colliding context variables but documents should not invoke both. They should choose the appropriate extension by invoking different tags.

PROPOSED:
Same as CURRENT with the following differences. If default behavior of an extension is desired, the extension tag does not need to be present in the document. Extensions providing alternative behavior cannot use colliding context variables. Their default behaviors must not conflict. If they do conflict only one of the extensions can be implied by an absence of a tag. The others must require that a tag be present as a kind of override.

For definition extension I see two minor downsides to PROPOSED. The first downside that comes to mind is consistency. Currently invoking functionality of any of the extensions requires asking for it in the template. That is pretty small difference. The second downside is performance related. Right now scanning the source for words happens when the `vocabulary` tag is first seen. This is simpler than waiting for first actual request for `words` or `has`. See related issue below.

Here is a related issue:
Suppose down the road someone, maybe us, needs to have a different definition of 'word'. Perhaps they need to include hyphens. Or look for specific phrases like "check in". So they create such an extension and give it a tag like `hyphenated_words` or `jargon`. I think the natural way to proceed would be to have the document include a tag like:
```
{% jargon system.common_phrases %}
...
{% if jargon.has('check in') %}
check in: submitting software changes to a source control system.
{% endif %}
```
For consistency with the above approach I suggest we change how we access words in the document. I think we should use `{{ vocabulary.words }}` instead of `{{ first_pass_output.words }}`. In the same spirit we should have `{{ vocabulary_ignore_case.words }}`.

I suggest we go with PROPOSED but change the context variable to `vocabulary` and `vocabulary_ignore_case`. This simplifies the documents with only a small constraint on future flexibility.